### ISSUE TITLE
[Feature request] Environment variables in duckdb CLI

### ISSUE THREAD

**Issue by mskyttner** [duckdb/duckdb#4038]
To facilitate making system specific settings in duckdb CLI, a feature allowing to make settings using values from system environment variables would be convenient.

For example in a CLI session when issuing initial PRAGMA statements for setting threads and memory limit or location of temp files, these settings could be made from system environment variables.

Another similar use case is connection strings, such as when using duckdb CLI with local object storage and setting connection information for an S3 backend. 

Consider running `duckdb -init statements.sql` with these contents:

```bash
load '/extensions/httpfs.duckdb_extension';
.shell echo "Number of CPUs: $(nproc --all)  # could be used for PRAGMA threads setting?"
set s3_endpoint=$MINIO_ENDPOINT;
set s3_access_key_id=$MINIO_USER;
set s3_secret_access_key=$MINIO_PASS;
set s3_use_ssl=false;
set s3_region='';
set s3_url_style='path';
.timer on
select count(*) from 's3://$MINIO_BUCKET/*.parquet';
```
If direct variable substitution is out of scope, could the environment variables for $MINIO_* that are set on the host be set through a `getenv` or similar function for picking up and using these in the duckdb CLI?

When packaging duckdb CLI, for example in a container, a mechanism to allow configuring settings through environment variables would be very useful.

#### Identity Disclosure:
 - Full Name: Markus Skyttner
 - Affiliation: KTH - Royal Instititute of Technology

**Comment by mskyttner** [comment]
These might be slightly related:

- #3855 
- #4021

**Comment by mskyttner** [comment]
I didn't know of [that](https://github.com/duckdb/duckdb/blob/master/tools/shell/shell.c) and will try it to see if it can be used to expose system variable values from the host to the CLI. One reply in this [SO post](https://stackoverflow.com/questions/27391557/how-to-make-sqliterc-commands-be-quiet) appears to view it as "not recommended" at least for sqlite when making PRAGMA settings, not sure why...

**Comment by mskyttner** [comment]
A utility function "getenv(envvarname)" that returned a string with the value for the named environment variable would be nice.

It could be used in the from clause ie

SELECT count(*) FROM getenv('S3_URI') || '/*.parquet'  

or when setting a pragma

PRAGMA threads=getenv('N_CPU')

and might be convenient as a pre-startup envsubst mechanism wouldn't have to be used to achieve variable substitution based on the .duckdbrc file?

**Comment by rupurt** [comment]
It would be great to be able to pass alternative `.duckdbrc` file paths via CLI.

e.g. in a nix environment

**Comment by Mause** [comment]
> It would be great to be able to pass alternative `.duckdbrc` file paths via CLI.
> 
> e.g. in a nix environment

You can use the existing `--init` flag for that purpose

**Comment by rupurt** [comment]
@Mause sweet! Thank you for that 🙌

**Comment by catkins** [comment]
So I came up with a bit of a workaround for not supporting env vars in the CLI.

I use [aws-vault](https://github.com/99designs/aws-vault) to assume role into various aws accounts/roles at work. aws-vault exec will run commands in a subshell with the STS env vars set for me, which is really handy for working with the aws cli or terraform.

With these two files: `init.sql` and `init.template.sql` and writing out a temporary file, we can work around this.

```sql
-- init.sql

-- evaluate template into a temporary sql file
.shell cat init.template.sql | envsubst > init-temp.sql

-- run the temporary sql file
.read init-temp.sql

-- clean up the temporary file
.shell rm init-temp.sql
```

```sql
-- init.template.sql

.print 'Preloading httpfs extension'
INSTALL httpfs;
LOAD httpfs;

.print 'Configuring AWS credentials from environment'
SET s3_region='${AWS_REGION}';
SET s3_access_key_id='${AWS_ACCESS_KEY_ID}';
SET s3_secret_access_key='${AWS_SECRET_ACCESS_KEY}';
SET s3_session_token='${AWS_SESSION_TOKEN}';
```

usage:

```sh
$ aws-vault some-profile exec -- duckdb -init init.sql
-- Loading resources from init.sql
Preloading httpfs extension
Configuring AWS credentials from environment
v0.7.1 b00b93f0b1
Enter ".help" for usage hints.
D
```

**Comment by catkins** [comment]
I just tested that @Mytherin 💡!

**Comment by mskyttner** [comment]
For S3 connectivity against a non-AWS server (such as minio), the `AWS_ENDPOINT_URL` would be convenient, see [this SO post](https://stackoverflow.com/questions/52494196/is-there-any-way-to-specify-endpoint-url-in-aws-cli-config-file) and [these docs](https://docs.aws.amazon.com/sdkref/latest/guide/feature-ss-endpoints.html).

I do not think that environment variable is known by #5419 (it doesn't seem to be aware of `AWS_ENDPOINT_URL`), but I believe it should correspond to `SET s3_endpoint='<domain>.<tld>:<port>';`

**Comment by snth** [comment]
I haven't tested the environment variables yet but I'm also planning to use it against Minio so I believe I will also need the `AWS_ENDPOINT_URL` as @mskyttner points out.

+1 for that.

**Comment by archiewood** [comment]
This would be great. 
Specifically if I want to deploy a duckdb to a production environment it would be great to also be able to install and load extensions using environment variables, which would enable things like:

```
EXTENSION_HTTPFS=enabled npm run dev
```

or in prod environments, setting that env var so that you could

```
npm run build
```

and the duckdb would use the extension

**Comment by odysseu** [comment]
would be great

**Comment by Mytherin** [comment]
Implemented in #9651
